### ISSUE TITLE
Example RedisClient fails on repeated GET operation with non-existing key

### ISSUE THREAD

**Issue by sunrenjie** [netty/netty#5184]
### Symptom

Run the client io.netty.example.redis.RedisClient against a redis server, issue the command "GET nonexisting" ("nonexisting" can be any key not present in the redis server) twice and exceptions are thrown (error messages at the bottom).

### My analysis
At the begining io.netty.handler.codec.MessageAggregator#decode(), currentMessage is not null as expected for the second GET command, because the derived class RedisBulkStringAggregator did not finish the first GET cleanly, which is again because FullBulkStringRedisMessage.NULL_INSTANCE, the representation of null string in response, cannot pass the test io.netty.handler.codec.MessageAggregator#acceptInboundMessage() in order to find its way into io.netty.handler.codec.MessageAggregator#decode() to finish up the aggregation process.

### My solution
My idea is to let GET operations for non-existing keys resemble those for existing keys. That is, to represent null string in null response, use a DefaultLastBulkStringRedisContent with empty buffer instead of FullBulkStringRedisMessage.NULL_INSTANCE.

Please feel free to comment, thanks!

### FULL Error messages

Note that "nonexisting" is a non-existing key in the redis server; user input will be **bold**.

Enter Redis commands (quit to end)
**get nonexisting**
(null)
**get nonexisting**
exceptionCaught: io.netty.handler.codec.DecoderException: io.netty.handler.codec.MessageAggregationException
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:99)
	at io.netty.channel.ChannelHandlerInvokerUtil.invokeChannelReadNow(ChannelHandlerInvokerUtil.java:83)
	at io.netty.channel.DefaultChannelHandlerInvoker.invokeChannelRead(DefaultChannelHandlerInvoker.java:154)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:354)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:145)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:277)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:264)
	at io.netty.channel.ChannelHandlerInvokerUtil.invokeChannelReadNow(ChannelHandlerInvokerUtil.java:83)
	at io.netty.channel.DefaultChannelHandlerInvoker.invokeChannelRead(DefaultChannelHandlerInvoker.java:154)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:354)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:145)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:1078)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:117)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:544)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:485)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:399)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:371)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:742)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:145)
	at java.lang.Thread.run(Thread.java:745)
Caused by: io.netty.handler.codec.MessageAggregationException
	at io.netty.handler.codec.MessageAggregator.decode(MessageAggregator.java:195)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:89)
	... 19 more
(null)

**Comment by jongyeol** [comment]
Oh.. I'll check it soon.

**Comment by Scottmitch** [comment]
Thanks @jongyeol !

**Comment by sunrenjie** [comment]
Hi,

there may be more problems regarding representing null strings that I discovered while trying to create a patch for it for fun (I'm basically a student on it :) ). In io.netty.handler.codec.MessageAggregator#decode(), we have this line:

`CompositeByteBuf content = (CompositeByteBuf) currentMessage.content();`

That means currentMessage, returned from beginAggregation() method of the derived classes, is assumed to have its ByteBuf derived from CompositeByteBuf (as we will do appendPartialContent() on content later). But io.netty.handler.codec.redis.RedisBulkStringAggregator#beginAggregation() may return FullBulkStringRedisMessage.NULL_INSTANCE for special null string. This means, if null response results in two messages going into decode() ( as I initially considered above), the conversion line will explode.

Here I suggest some possibilities on dealing with the special null response strings:
1) Decode it into one message. Change MessageAggregator to allow this special non-chunked message.
2) Decode it into two messages. While EmptyByteBuf is still used to represent null strings, change decode() to not force use of CompositeByteBuf in currentMessage.content(). That may involve changing the ByteBufHolder interface to allow its underlying ByteBuf to be assigned.
3) Decode it into two messages. Use ordinary zero-length ByteBuf to represent null strings. There will be no way for the downstream handlers to tell the differences between null strings and zero-length string values.

Please feel free to correct me.

**Comment by sunrenjie** [comment]
Some nits on command line I/O handling of empty string values that I'd rather not open a new issue. @jongyeol @Scottmitch 

Below both redis-cli and RedisClient are running against the same redis server.

With redis-cli:
$ redis-cli
127.0.0.1:6379> set a ""
OK
127.0.0.1:6379> get a
""

With example RedisClient:
Enter Redis commands (quit to end)
get a

set b ''
OK
get b
''

Again with redis-cli:
$ redis-cli
127.0.0.1:6379> get b
"''"

Thanks!

**Comment by jongyeol** [comment]
@sunrenjie , First of all, thank you for reporting this. And, your opinions are right. I reviewed codes and I realized we don't need to seperate header and contents for null or empty bulk string, and pushed the PR #5193 for this. And, about second one, I have no idea for setting empty bulk string and I think this is not urgent because this is "an example usage for `codec-redis`". So, if you have some idea, how about you create a PR for it?
